File: Src/Engine/Editor/GUI/Drag/DragHandlers.cpp
#include "DragHandlers.h"

namespace SE::Editor
{
	DragDropEffect DragHandlers::OnDragEnter(DragData* data)
	{
		DragDropEffect effect = DragDropEffect::None;

		for (int i = 0; i < Count(); i++)
		{
			DragHelperBase* dragHelper = this->operator[](i);
			if (dragHelper->OnDragEnter(data))
			{
				effect = dragHelper->GetEffect();
			}
		}

		return effect;
	}

	void DragHandlers::OnDragLeave()
	{
		for (int i = 0; i < Count(); i++)
		{
			this[i].OnDragLeave();
		}
	}

	void DragHandlers::OnDragDrop(DragEventArgs* dragEventArgs)
	{
		for (int i = 0; i < Count(); i++)
		{
			this[i].OnDragDrop(dragEventArgs);
		}
	}

	bool DragHandlers::GetHasValidDrag()
	{
		for (int i = 0; i < Count(); i++)
		{
			DragHelperBase* dragHelper = this->operator[](i);
			if (dragHelper->GetHasValidDrag())
			{
				return true;
			}
		}
		return false;
	}

	DragHelperBase* DragHandlers::WithValidDrag()
	{
		if (IsEmpty())
		{
			return nullptr;
		}

		for (int i = 0; i < Count(); i++)
		{
			DragHelperBase* dragHelper = this->operator[](i);
			if (dragHelper->GetHasValidDrag())
			{
				return dragHelper;
			}
		}

		return nullptr;
	}

	DragDropEffect DragHandlers::GetEffect()
	{
		for (int i = 0; i < Count(); i++)
		{
			DragHelperBase* dragHelper = this->operator[](i);
			if (dragHelper->GetHasValidDrag())
			{
				return dragHelper->GetEffect();
			}
		}

		return DragDropEffect::None;
	}
}

File: Src/Engine/Editor/GUI/Drag/DragHandlers.cs
using System.Collections.Generic;

namespace SE.Editor.GUI
{
    public sealed class DragHandlers : List<DragHelperBase>
    {
        public DragDropEffect OnDragEnter(DragData data)
        {
            DragDropEffect result = DragDropEffect.None;
            foreach (DragHelperBase helper in this)
            {
                if (helper.OnDragEnter(data) && result == DragDropEffect.None)
                    result = helper.Effect;
            }

            return result;
        }

        public void OnDragLeave()
        {
            foreach (DragHelperBase helper in this)
            {
                helper.OnDragLeave();
            }
        }

        public void OnDragDrop(DragEventArgs dragEventArgs)
        {
            foreach (DragHelperBase helper in this)
            {
                helper.OnDragDrop(dragEventArgs);
            }
        }

        public bool HasValidDrag
        {
            get
            {
                return WithValidDrag() != null;
            }
        }

        public DragHelperBase? WithValidDrag()
        {
            foreach (DragHelperBase helper in this)
            {
                if (helper.HasValidDrag)
                    return helper;
            }

            return null;
        }

        public DragDropEffect Effect => WithValidDrag()?.Effect ?? DragDropEffect.None;
    }
}
